R/step_adjust_latency.R

```diff
@@ -0,0 +1,462 @@
+#'   forecast at every location, we need to guarantee data at every location by
+#'   using the largest latency across every location; this means taking
+#'   `max_time_value` to be the minimum of the `max_time_value`s for each set of
+#'   key values (so the earliest date).  If `NULL` or an empty character vector,
+#'   it will take the maximum across all values, irrespective of any keys.
+#'
+#'   Note that this is a separate concern from different latencies across
+#'   different *data columns*, which is only handled by the choice of `method`.
+#' @param fixed_latency either a positive integer, or a labeled positive integer
+#'   vector. Cannot be set at the same time as `fixed_forecast_date`. If
+#'   non-`NULL`, the amount to offset the ahead or lag by. If a single integer,
+#'   this is used for all columns; if a labeled vector, the labels must
+#'   correspond to the base column names (before lags/aheads).  If `NULL`, the
+#'   latency is the distance between the `epi_df`'s `max_time_value` and the `forecast_date`.
+#' @param fixed_forecast_date either a date of the same kind used in the
+#'   `epi_df`, or `NULL`. Exclusive with `fixed_latency`. If a date, it gives
+#'   the date from which the forecast is actually occurring. If `NULL`, the
+#'   `forecast_date` is determined either via the `fixed_latency`, or is set to
+#'   the `epi_df`'s `as_of` value if `fixed_latency` is also `NULL`.
+#' @param role For model terms created by this step, what analysis role should
+#'   they be assigned? `lag` is a predictor while `ahead` is an outcome.  It
+#'   should be correctly inferred and not need setting
+#' @template step-return
+#' @inheritParams recipes::step_lag
+#'
+#'
+#' @family row operation steps
+#' @rdname step_adjust_latency
+#' @export
+#' @examples
+#' jhu <- case_death_rate_subset %>%
+#'   dplyr::filter(time_value > "2021-11-01", geo_value %in% c("ak", "ca", "ny"))
+#' # setting the `as_of` to something realistic
+#' attributes(jhu)$metadata$as_of <- max(jhu$time_value) + 3
+#'
+#' r <- epi_recipe(case_death_rate_subset) %>%
+#'   step_adjust_latency(method = "extend_ahead") %>%
+#'   step_epi_ahead(death_rate, ahead = 7) %>%
+#'   step_epi_lag(death_rate, lag = c(0, 7, 14))
+#' r
+#'
+#' jhu_fit <- epi_workflow() %>%
+#'   add_epi_recipe(r) %>%
+#'   add_model(linear_reg()) %>%
+#'   fit(data = jhu)
+#' jhu_fit
+#'
+#' @importFrom recipes detect_step
+#' @importFrom rlang enquos is_empty
+#' @importFrom dplyr tribble n
+step_adjust_latency <-
+  function(recipe,
+           ...,
+           role = NA,
+           trained = FALSE,
+           method = c(
+             "extend_ahead",
+             "locf",
+             "extend_lags"
+           ),
+           epi_keys_checked = c("geo_value"),
+           fixed_latency = NULL,
+           fixed_forecast_date = NULL,
+           skip = FALSE,
+           columns = NULL,
+           id = recipes::rand_id("adjust_latency")) {
```

Comment:
A few things here:
1. Unlike `{recipes}`, I don't think we should provide arguments that the user shouldn't touch. So `trained` and `columns` should be removed here and set explicitly to these values at the bottom of this function.
2. `role`. What is this used for? It's a bit ambiguous to me. You are using this step to potentially adjust outcomes or predictors depending on other arguments. So I think it should be removed unless I'm missing something.
3. `skip`. This is also potentially problematic. In standard usage, `FALSE` means that this step is evaluated at both train time and test time. I suspect that's what we want always. `TRUE` means it will NOT be evaluated at test time. This could be very strange, I think.
4. `rand_id` is imported.
5. are `...` intended to be used with tidy-selection? I think they should be empty, right? Unless `method = "locf"` maybe? In which case `role` may be needed?
6. I'm not sure about `epi_keys_checked`. What if the epi_df has other keys?

So my suggestion follows this model, but has many implications in the function body. And I'm not sure my understanding is correct.

```suggestion
  function(recipe,
                 method = c("extend_ahead", "locf", "extend_lags"),
           epi_keys_checked = c("geo_value"),
           fixed_latency = NULL,
           fixed_forecast_date = NULL,
           id = rand_id("adjust_latency")) {
```

Reply:
Most of that tracks, though 
5. `...` are for tidy selection, if for whatever reason you don't want to apply it to every column, or if you want to apply it to specifically non-raw columns.
6. `epi_keys_checked` should probably default to all epi_keys besides time, so I'll switch it to that.